### PR TITLE
Add operation to save image snapshots to local disk.

### PR DESCRIPTION
**Pull request by PeterJohnson**
The period and path prefix and suffix can be set, and the operation enabled
and disabled at will (this is particularly useful in combination with PR #572).

### CODE REVIEW

File: core/src/main/java/edu/wpi/grip/core/operations/composite/SaveImageOperation.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+                    // Copy the image data into a pre-allocated buffer, growing it if necessary
+                    bufferSize = imagePointer.limit();
+                    if (bufferSize > buffer.length) buffer = new byte[imagePointer.limit()];
+                    imagePointer.get(buffer, 0, bufferSize);
+                    prefix = this.prefix;
+                    suffix = this.suffix;
+                }
+            } catch (InterruptedException e) {
+                break;
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/SaveImageOperation.java]: You need to reset the interrupt flag here.
`Thread.interrupt()`

File: core/src/main/java/edu/wpi/grip/core/operations/composite/SaveImageOperation.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            }
+
+            // Write the file
+            try {
+                FileOutputStream out = new FileOutputStream(prefix + format.format(new Date()) + suffix);
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/SaveImageOperation.java]: Try with resource would be more appropriate here.
https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

File: core/src/main/java/edu/wpi/grip/core/operations/composite/SaveImageOperation.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            return;
+        }
+
+        if (input.empty()) {
+            throw new IllegalArgumentException("Input image must not be empty");
+        }
+
+        //if (period.doubleValue() < 0.03) {
+        //    throw new IllegalArgumentException("Delay must be more than 30 ms");
+        //}
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/SaveImageOperation.java]: Why is this not a constraint?
Should this comment be deleted?

**Comment by JLLeitschuh** [comment]
Thanks for this. This is not something I thought people would want. I like the idea though.

File: core/src/main/java/edu/wpi/grip/core/operations/composite/SaveImageOperation.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+ */
+public class SaveImageOperation implements Operation {
+
+    private final Logger logger = Logger.getLogger(SaveImageOperation.class.getName());
+
+    private final Object imageLock = new Object();
+    private final BytePointer imagePointer = new BytePointer();
+    private Optional<Thread> saveThread = Optional.empty();
+    private volatile boolean connected = false;
+    private long startTime = System.nanoTime();
```
Comment (JLLeitschuh) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/SaveImageOperation.java]: Guava has the [stopwatch](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/base/Stopwatch.html) that may be more apropreate here. Up to you though.

Comment (SamCarlberg) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/SaveImageOperation.java]: I don't think this even needs to be timed; it'll only run when the pipeline gets fired anyway.

**Comment by PeterJohnson** [comment]
This is very useful for testing and debugging at competition where you don't have realtime access because you can load up the saved images to play them back through the pipeline and adjust thresholds etc.  I'll fix up based on the comments so far.

File: core/src/main/java/edu/wpi/grip/core/operations/composite/SaveImageOperation.java (hunk 4 above)
Comment (PeterJohnson) [inline core/src/main/java/edu/wpi/grip/core/operations/composite/SaveImageOperation.java]: Well, it's nice to not necessarily save *every* frame all the time.  For example, my team saves images twice a second during a match but switches to saving every image when actually auto-aiming.

**Comment by PeterJohnson** [comment]
I'm debating adding a directory chooser for the output directory instead of or in addition to the prefix.  However, I'm not sure if adding those sorts of controls is frowned upon in operations.

**Comment by JLLeitschuh** [comment]
No, a directory chooser would be a good control to add.

**Pull request closed without merging**